api/cache.go

```diff
@@ -0,0 +1,123 @@
+package api
+
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"github.com/patrickmn/go-cache"
+	"github.com/trustwallet/blockatlas/pkg/logger"
+	"net/http"
+	"time"
+)
+
+var (
+	memoryCache *cache.Cache
+)
+
+func init() {
+	memoryCache = cache.New(5*time.Minute, 5*time.Minute)
+}
+
+type cacheResponse struct {
+	Status int
+	Header http.Header
+	Data   []byte
+}
+
+type cachedWriter struct {
+	gin.ResponseWriter
+	status  int
+	written bool
+	expire  time.Duration
+	key     string
+}
+
+var _ gin.ResponseWriter = &cachedWriter{}
+
+func newCachedWriter(expire time.Duration, writer gin.ResponseWriter, key string) *cachedWriter {
+	return &cachedWriter{writer, 0, false, expire, key}
+}
+
+func (w *cachedWriter) WriteHeader(code int) {
+	w.status = code
+	w.written = true
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func (w *cachedWriter) Status() int {
+	return w.ResponseWriter.Status()
+}
+
+func (w *cachedWriter) Written() bool {
+	return w.ResponseWriter.Written()
+}
+
+func getCacheResponse(key string) (*cacheResponse, error) {
+	mc, ok := memoryCache.Get(key)
+	if !ok {
+		return nil, fmt.Errorf("gin-cache: invalid cache key %s", key)
+	}
+
+	tempCache, ok := mc.(cacheResponse)
+	if !ok {
+		return nil, fmt.Errorf("gin-cache: invalid cache object %s", key)
+	}
+	return &tempCache, nil
+}
+
+func (w *cachedWriter) Write(data []byte) (int, error) {
+	ret, err := w.ResponseWriter.Write(data)
+	if err != nil {
+		return 0, err
+	}
+	if w.Status() < 300 {
+		val := cacheResponse{
+			w.Status(),
+			w.Header(),
+			data,
+		}
+		memoryCache.Set(w.key, val, w.expire)
+	}
+	return ret, nil
+}
+
+func (w *cachedWriter) WriteString(data string) (n int, err error) {
+	ret, err := w.ResponseWriter.WriteString(data)
+	if err == nil && w.Status() < 300 {
+		val := cacheResponse{
+			w.Status(),
+			w.Header(),
+			[]byte(data),
+		}
+		memoryCache.Set(w.key, val, w.expire)
+	}
+	return ret, err
+}
+
+func CacheMiddleware(minLength int, expiration time.Duration, handle gin.HandlerFunc) gin.HandlerFunc {
```

Comment:
What is minLenght?

Reply:
I added a parameter minLength. It's used in case the API returns empty docs, like { "docs": [] }, we don't gonna cache that

Reply:
I would suggest to remove `minLength` and make `validators()` to fail if it wasn't able to fetch data. in this way it would be very specific. 

Unless better way to handle it?

Reply:
it's better!!! done!!